InvalidFieldException using __defaults__ and environments-preview

**Describe the bug**
I have configured the `[environments-preview.names]` section in my `pants.toml`. Additionally I have applied the default to skip mypy for all my tests in the top level BUILD file where I have also defined my `docker_environment`. When I now run any pants command I get the below error:

```
➜ pants --version
13:49:46.90 [ERROR] 1 Exception encountered:

MappingError: Failed to parse ./BUILD:
InvalidFieldException: Unrecognized field `skip_mypy` for target python_tests. Valid fields are: batch_compatibility_tag, dependencies, description, environment, extra_env_vars, interpreter_constraints, overrides, resolve, run_goal_use_sandbox, runtime_package_dependencies, skip_tests, sources, tags, timeout, xdist_concurrency.


Use --print-stacktrace for more error details and/or -ldebug for more logs. 
See https://www.pantsbuild.org/v2.16/docs/troubleshooting for common issues.
Consider reaching out for help: https://www.pantsbuild.org/v2.16/docs/getting-help
```

When I remove the `[environments-preview.names]` section the commands work again:
```
➜ pants --version
13:53:34.74 [INFO] Initializing scheduler...
13:53:38.14 [INFO] Scheduler initialized.
2.16.0.dev5
```

A clear and concise description of the bug.

**Pants version**
2.16.0.dev5, we are using the new scie-pants binary in version 0.3.2

**OS**
MacOS

**Additional info**
I have a reproducible example on the branch here https://github.com/psontag/pexample/tree/pants-reproduce-field-issue
Just run any pants command to trigger it.

--- a/docs/markdown/Using Pants/concepts/targets.md
+++ b/docs/markdown/Using Pants/concepts/targets.md
@@ -167,7 +167,7 @@ Use the `all` keyword argument to provide default field values that should apply
 
 The `extend=True` keyword argument allows to add to any existing default field values set by a previous `__defaults__` call rather than replacing them.
 
-Default fields and values are validated against their target types, except when provided using the `all` keyword, in which case only values for fields applicable to each target are validated.
+Default fields and values are validated against their target types, except when provided using the `all` keyword, in which case only values for fields applicable to each target are validated. Use `ignore_unknown_fields=True` to ignore invalid fields.
 
 This means, that it is legal to provide a default value for `all` targets, even if it is only a subset of targets that actually supports that particular field.
 
@@ -199,6 +199,21 @@ To reset any modified defaults, simply override with the empty dict:
 ```python src/example/nodefaults/BUILD
     __defaults__(all={})
 ```
+
+Supporting optional plugin fields
+---------------------------------
+
+Normally Pants presents an error message when attempting to provide a default value for a field that doesn't exist for the target. However, some fields comes from plugins, and to support disabling a plugin without having to remove any default values referencing any plugin fields it was providing, there is a `ignore_unknown_fields` option to use:
+
+```python example/BUILD
+    __defaults__(
+      {
+        # Defaults...
+      },
+      ignore_unknown_fields=True,
+    )
+```
+
 
 Target generation
 =================

--- a/src/python/pants/engine/internals/build_files.py
+++ b/src/python/pants/engine/internals/build_files.py
@@ -160,6 +160,7 @@ async def ensure_address_family(request: OptionalAddressFamily) -> AddressFamily
 @rule(desc="Search for addresses in BUILD files")
 async def parse_address_family(
     parser: Parser,
+    bootstrap_status: BootstrapStatus,
     build_file_options: BuildFileOptions,
     prelude_symbols: BuildFilePreludeSymbols,
     directory: AddressFamilyDir,
@@ -203,7 +204,12 @@ async def parse_address_family(
     )
     address_maps = [
         AddressMap.parse(
-            fc.path, fc.content.decode(), parser, prelude_symbols, defaults_parser_state
+            fc.path,
+            fc.content.decode(),
+            parser,
+            prelude_symbols,
+            bootstrap_status.in_progress,
+            defaults_parser_state,
         )
         for fc in digest_contents
     ]

--- a/src/python/pants/engine/internals/build_files_test.py
+++ b/src/python/pants/engine/internals/build_files_test.py
@@ -39,6 +39,7 @@ from pants.engine.target import (
     Target,
 )
 from pants.engine.unions import UnionMembership
+from pants.init.bootstrap_scheduler import BootstrapStatus
 from pants.testutil.rule_runner import (
     MockGet,
     QueryRule,
@@ -60,6 +61,7 @@ def test_parse_address_family_empty() -> None:
                 object_aliases=BuildFileAliases(),
                 ignore_unrecognized_symbols=False,
             ),
+            BootstrapStatus(in_progress=False),
             BuildFileOptions(("BUILD",)),
             BuildFilePreludeSymbols(FrozenDict()),
             AddressFamilyDir("/dev/null"),
@@ -493,3 +495,25 @@ def test_macro_undefined_symbol_bootstrap() -> None:
     # Parse the root BUILD file.
     address_family = rule_runner.request(AddressFamily, [AddressFamilyDir("")])
     assert not address_family.name_to_target_adaptors
+
+
+def test_default_plugin_field_bootstrap() -> None:
+    # Tests that an unknown field in `__defaults__` is ignored while bootstrapping.
+    rule_runner = RuleRunner(
+        rules=[QueryRule(AddressFamily, [AddressFamilyDir])],
+        target_types=[MockTgt],
+        is_bootstrap=True,
+    )
+    rule_runner.write_files(
+        {
+            "BUILD": dedent(
+                """
+                __defaults__({mock_tgt: dict(presumably_plugin_field="default", tags=["ok"])})
+                """
+            ),
+        }
+    )
+
+    # Parse the root BUILD file.
+    address_family = rule_runner.request(AddressFamily, [AddressFamilyDir("")])
+    assert dict(tags=("ok",)) == dict(address_family.defaults["mock_tgt"])

--- a/src/python/pants/engine/internals/defaults.py
+++ b/src/python/pants/engine/internals/defaults.py
@@ -127,6 +127,7 @@ class BuildFileDefaultsParserState:
         *args: SetDefaultsT,
         all: SetDefaultsValueT | None = None,
         extend: bool = False,
+        ignore_unknown_fields: bool = False,
         **kwargs,
     ) -> None:
         defaults: dict[str, dict[str, Any]] = (
@@ -141,7 +142,7 @@ class BuildFileDefaultsParserState:
             )
 
         for arg in args:
-            self._process_defaults(defaults, arg)
+            self._process_defaults(defaults, arg, ignore_unknown_fields=ignore_unknown_fields)
 
         # Update with new defaults, dropping targets without any default values.
         for tgt, default in defaults.items():

--- a/src/python/pants/engine/internals/mapper.py
+++ b/src/python/pants/engine/internals/mapper.py
@@ -39,6 +39,7 @@ class AddressMap:
         build_file_content: str,
         parser: Parser,
         extra_symbols: BuildFilePreludeSymbols,
+        is_bootstrap: bool,
         defaults: BuildFileDefaultsParserState,
     ) -> AddressMap:
         """Parses a source for targets.
@@ -47,7 +48,9 @@ class AddressMap:
         the same namespace but from a separate source are left as unresolved pointers.
         """
         try:
-            target_adaptors = parser.parse(filepath, build_file_content, extra_symbols, defaults)
+            target_adaptors = parser.parse(
+                filepath, build_file_content, extra_symbols, is_bootstrap, defaults
+            )
         except Exception as e:
             raise MappingError(f"Failed to parse ./{filepath}:\n{e}")
         return cls.create(filepath, target_adaptors)

--- a/src/python/pants/engine/internals/mapper_test.py
+++ b/src/python/pants/engine/internals/mapper_test.py
@@ -40,6 +40,7 @@ def parse_address_map(build_file: str, *, ignore_unrecognized_symbols: bool = Fa
         build_file,
         parser,
         BuildFilePreludeSymbols(FrozenDict()),
+        False,
         BuildFileDefaultsParserState.create(
             "", BuildFileDefaults({}), RegisteredTargetTypes({}), UnionMembership({})
         ),

--- a/src/python/pants/engine/internals/parser.py
+++ b/src/python/pants/engine/internals/parser.py
@@ -41,8 +41,12 @@ class ParseState(threading.local):
         self._defaults: BuildFileDefaultsParserState | None = None
         self._rel_path: str | None = None
         self._target_adapters: list[TargetAdaptor] = []
+        self._is_bootstrap: bool | None = None
 
-    def reset(self, rel_path: str, defaults: BuildFileDefaultsParserState) -> None:
+    def reset(
+        self, rel_path: str, is_bootstrap: bool, defaults: BuildFileDefaultsParserState
+    ) -> None:
+        self._is_bootstrap = is_bootstrap
         self._defaults = defaults
         self._rel_path = rel_path
         self._target_adapters.clear()
@@ -72,8 +76,21 @@ class ParseState(threading.local):
             )
         return self._defaults
 
-    def set_defaults(self, *args: SetDefaultsT, **kwargs) -> None:
-        self.defaults.set_defaults(*args, **kwargs)
+    @property
+    def is_bootstrap(self) -> bool:
+        if self._is_bootstrap is None:
+            raise AssertionError(
+                "Internal error in Pants. Please file a bug report at "
+                "https://github.com/pantsbuild/pants/issues/new"
+            )
+        return self._is_bootstrap
+
+    def set_defaults(
+        self, *args: SetDefaultsT, ignore_unknown_fields: bool = False, **kwargs
+    ) -> None:
+        self.defaults.set_defaults(
+            *args, ignore_unknown_fields=self.is_bootstrap or ignore_unknown_fields, **kwargs
+        )
 
 
 class Parser:
@@ -153,9 +170,12 @@ class Parser:
         filepath: str,
         build_file_content: str,
         extra_symbols: BuildFilePreludeSymbols,
+        is_bootstrap: bool,
         defaults: BuildFileDefaultsParserState,
     ) -> list[TargetAdaptor]:
-        self._parse_state.reset(rel_path=os.path.dirname(filepath), defaults=defaults)
+        self._parse_state.reset(
+            rel_path=os.path.dirname(filepath), is_bootstrap=is_bootstrap, defaults=defaults
+        )
 
         global_symbols = {**self._symbols, **extra_symbols.symbols}
 
@@ -174,7 +194,11 @@ class Parser:
                     defined_symbols.add(bad_symbol)
 
                     global_symbols[bad_symbol] = _unrecognized_symbol_func
-                    self._parse_state.reset(rel_path=os.path.dirname(filepath), defaults=defaults)
+                    self._parse_state.reset(
+                        rel_path=os.path.dirname(filepath),
+                        is_bootstrap=is_bootstrap,
+                        defaults=defaults,
+                    )
                     continue
                 break
 

--- a/src/python/pants/engine/internals/parser_test.py
+++ b/src/python/pants/engine/internals/parser_test.py
@@ -39,6 +39,7 @@ def test_imports_banned(defaults_parser_state: BuildFileDefaultsParserState) -> 
             "dir/BUILD",
             "\nx = 'hello'\n\nimport os\n",
             BuildFilePreludeSymbols(FrozenDict()),
+            False,
             defaults_parser_state,
         )
     assert "Import used in dir/BUILD at line 4" in str(exc.value)
@@ -64,6 +65,7 @@ def test_unrecognized_symbol(defaults_parser_state: BuildFileDefaultsParserState
                 "dir/BUILD",
                 "fake",
                 prelude_symbols,
+                False,
                 defaults_parser_state,
             )
         assert str(exc.value) == (
@@ -86,6 +88,7 @@ def test_unrecognized_symbol(defaults_parser_state: BuildFileDefaultsParserState
                 "dir/BUILD",
                 "fake",
                 prelude_symbols,
+                False,
                 defaults_parser_state,
             )
 

